src/lib.rs

```diff
@@ -1,3 +1,162 @@
+// "instructions", if we decide to go that route.
+pub mod tokens {
+
+    // It is possible to delay all string manipulation for later,
+    // and we can simply store the slices into original full source
+    // string.
+    //
+    // We can keep doing that even for Node<'a>, provided the strings remain untouched.
+    // If something needs to be changed, we can create a special Node for that.
+    #[derive(Debug)]
+    pub enum TokenRef<'a> {
+        Text(&'a str),
+    }
+
+    impl<'a> TokenRef<'a> {
+        // Not used "into", because Gankro criticises using into for anything more than
+        // moving data around. Not used into_owned, because we don't implement ToOwned trait.
+        // So the only logical name remains `into_token`.
+        pub fn into_token(self) -> Token {
+            match self {
+                TokenRef::Text(v) => Token::Text(v.into()),
+            }
+        }
+    }
+
+    // This will be used when we need to carry token lifetime longer than original
+    // source string, for example, in error messages.
+    pub enum Token {
+        Text(String),
+    }
+
+    // Not pub, to make API more convenient.
+    mod lexing {
+        use Position;
+        use tokens::TokenRef;
+
+
+
+        /// TokenRef wrapper for `Lexer` that additionaly has position.
+        #[derive(Debug)]
+        pub struct ItemRef<'t> {
+            pub token: TokenRef<'t>,
+            pub position: Position,
+        }
+
+
+
+        // TBD simple lexer options (delimiters, whitespace, etc).
+        #[derive(Copy, Clone)]
+        pub struct Options;
+
+        impl Options {
+            pub fn default() -> Options { Options }
+        }
+
+
+
+        // I will be refering to 't as template lifetime, 'i as iteration lifetime.
+        // This lexer should be reusable between the `tokenize` calls.
+        // In addition to this I had `LexingEnvironment`, but it turned out to be redundant.
+        pub struct Lexer;
+
+        impl Lexer {
+            // It's responsibility of someone else to take operators from extensions,
+            // resolve any conflicts and compile final "operators" list.
+            //
+            // It looks like Lexer does not care if they are unary or binary, that will
+            // become important in parser.
+            //
+            // Funny note: I found that "=" is considered neither unary nor binary ;)
+            pub fn new(options: Options, operators: Vec<&'static str>) -> Lexer {
+                // Here we will create patterns (I called them matchers), and
+                // store them in Lexer
+                Lexer
+            }
+
+            // twig-rust: https://github.com/colin-kiegel/twig-rust/blob/master/src/lexer/mod.rs#L64
+            // twig-rs: https://github.com/Nercury/twig-rs/blob/master/src/tokens/lexer/mod.rs#L40
+            //
+            // I think it is possible to avoid the Template::Raw in lexer API.
```

Comment:
Yes, Template::Raw turned out to be quite close to useless. I was just giving everything names. On the other hand, it didn't hurt - so I kept it. :-)